src/main/java/org/dataone/hashstore/hashfs/HashFileStore.java

```diff
@@ -127,91 +121,124 @@ public HashAddress putObject(InputStream object, String pid, String additionalAl
         // If file (pid hash) exists, reject request immediately
         if (objHashAddress.exists()) {
+            logHashFileStore.error("HashFileStore.putObject - File already exists for pid: " + pid
+                    + ". Object address: " + objAbsolutePathString);
             throw new FileAlreadyExistsException("File already exists for pid: " + pid);
         }
 
         // Generate tmp file and write to it
-        File tmpFile = this.hashUtil.generateTmpFile("tmp", this.tmpFileDirectory.toFile());
-        Map<String, String> hexDigests = this.hashUtil.writeToTmpFileAndGenerateChecksums(tmpFile, object,
+        logHashFileStore.debug("HashFileStore.putObject - Generating tmpFile");
+        File tmpFile = this.generateTmpFile("tmp", this.tmpFileDirectory);
+        Map<String, String> hexDigests = this.writeToTmpFileAndGenerateChecksums(tmpFile, object,
                 additionalAlgorithm);
 
         // Validate object if checksum and checksum algorithm is passed
         if (checksumAlgorithm != null && checksum != null) {
+            logHashFileStore.info("HashFileStore.putObject - Validating object");
             String digestFromHexDigests = hexDigests.get(checksumAlgorithm);
             if (!checksum.equals(digestFromHexDigests)) {
-                tmpFile.delete();
-                // TODO: Log failure - include signature values
+                // Delete tmp File
+                boolean deleteStatus = tmpFile.delete();
+                if (!deleteStatus) {
+                    throw new IOException("Object cannot be validated and failed to delete tmpFile: " + tmpFile);
+                }
+                logHashFileStore.error(
+                        "HashFileStore.putObject - Checksum supplied does not equal to the calculated hex digest: "
+                                + digestFromHexDigests + ". Checksum provided: " + checksum + ". Deleting tmpFile: "
+                                + tmpFile);
                 throw new IllegalArgumentException(
                         "Checksum supplied does not equal to the calculated hex digest: " + digestFromHexDigests
-                                + ". Checksum provided: " + checksum + ". Deleting tmpFile: " + tmpFile.toString());
+                                + ". Checksum provided: " + checksum + ". Deleting tmpFile: " + tmpFile);
             }
         }
 
         // Move object
-        boolean isNotDuplicate = this.hashUtil.move(tmpFile, objHashAddress);
+        logHashFileStore.debug("HashFileStore.putObject - Moving object: " + tmpFile.toString() + ". Destination: "
+                + objAbsolutePathString);
+        boolean isNotDuplicate = this.move(tmpFile, objHashAddress);
+        logHashFileStore
+                .info("HashFileStore.putObject - Move object success, permanent address: " + objAbsolutePathString);
         if (!isNotDuplicate) {
-            tmpFile.delete();
+            boolean deleteStatus = tmpFile.delete();
```

Comment:
Why should we delete the temp file? The move method already renames it, right?

Reply:
Discussion for this in thread above.